File: GUILib/src/space/kiibou/gui/GraphicsElement.java
package space.kiibou.gui;

import processing.core.PImage;
import space.kiibou.GApplet;
import space.kiibou.event.MouseEvent;
import space.kiibou.event.MouseEventListener;
import space.kiibou.event.MouseEventOption;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;

public abstract class GraphicsElement extends Rectangle implements MouseEventListener {
    private static int gID = 0;

    private final GApplet app;

    private boolean hidden;
    private final int scale;
    private final List<GraphicsElement> children;
    private GraphicsElement parent;

    private boolean active;
    private final MouseOptionMap mouseOptionMap;

    private int hierarchyDepth;
    private final int id;

    private boolean preInitialized;
    private boolean initialized;
    private boolean postInitialized;

    public GraphicsElement(GApplet app, int x, int y, int width, int height, int scale) {
        super(x, y, width, height);
        this.app = app;
        this.hidden = false;
        this.scale = scale;
        children = Collections.synchronizedList(new ArrayList<>());
        mouseOptionMap = new MouseOptionMap();
        active = true;
        hierarchyDepth = 0;
        id = nextID();
        preInitialized = false;
        initialized = false;
        postInitialized = false;
    }

    static Function<Rectangle, TileRenderer> tilemapRenderFactory(PImage image) {
        return (rectangle -> tileRenderFactory(rectangle.getX(), rectangle.getY(), rectangle.getX() + rectangle.getWidth(), rectangle.getY() + rectangle.getHeight(), image));
    }

    private static TileRenderer tileRenderFactory(int u1, int v1, int u2, int v2, PImage tilemap) {
        return ((graphics, rectangle) -> graphics.image(tilemap, rectangle.getX(), rectangle.getY(), rectangle.getWidth(), rectangle.getHeight(), u1, v1, u2, v2));
    }

    private static int nextID() {
        return gID++;
    }

    @Override
    public void registerCallback(MouseEventOption option, Consumer<MouseEvent> callback) {
        MouseEventListener.super.registerCallback(option, callback);
        app.registerMethod("mouseEvent", this);
    }

    @Override
    public void unregisterCallback(MouseEventOption option) {
        MouseEventListener.super.unregisterCallback(option);
        if (mouseOptionMap.isEmpty()) {
            app.unregisterMethod("mouseEvent", this);
        }
    }

    public final void preInit() {
        preInitImpl();
        preInitialized = true;
        getChildren().forEach(GraphicsElement::preInit);
    }

    /**
     * Called before children are pre initialized
     */
    protected abstract void preInitImpl();

    public final void init() {
        getChildren().forEach(GraphicsElement::init);
        initImpl();
        initialized = true;
    }

    /**
     * Called after children are initialized
     */
    protected abstract void initImpl();

    public final void postInit() {
        getChildren().forEach(GraphicsElement::postInit);
        postInitImpl();
        postInitialized = true;
    }

    /**
     * Called before children are post initialized
     */
    protected abstract void postInitImpl();

    public boolean isPreInitialized() {
        return preInitialized;
    }

    public boolean isInitialized() {
        return initialized;
    }

    public boolean isPostInitialized() {
        return postInitialized;
    }

    public final void draw() {
        if (!hidden) {
            drawImpl();
            getChildren().forEach(GraphicsElement::draw);
        }
    }

    protected abstract void drawImpl();


    public GraphicsElement moveTo(int x, int y) {
        int dx = x - getX();
        int dy = y - getY();
        move(dx, dy);
        return this;
    }

    public GraphicsElement move(int x, int y) {
        getChildren().forEach(child -> child.move(x, y));
        setX(getX() + x);
        setY(getY() + y);
        return this;
    }

    public GraphicsElement resize(int width, int height) {
        setWidth(width);
        setHeight(height);
        return this;
    }

    public GraphicsElement resizeUnscaled(int width, int height) {
        setWidth(width * scale);
        setHeight(height * scale);
        return this;
    }

    public int getUnscaledWidth() {
        return getWidth() / getScale();
    }

    public int getUnscaledHeight() {
        return getHeight() / getScale();
    }

    public int getScale() {
        return scale;
    }

    public List<GraphicsElement> getChildren() {
        return children;
    }

    public void addChild(GraphicsElement element) {
        addChild(children.size(), element);
    }

    public void addChild(int index, GraphicsElement element) {
        if (element.getParent() == null) {
            element.setParent(this);
            getChildren().add(index, element);
            element.hierarchyDepth = hierarchyDepth + 1;
        } else {
            throw new IllegalArgumentException("The passed GraphicsElement was is already child of another GraphicsElement.");
        }
    }

    public GraphicsElement getChild(int index) {
        return children.get(index);
    }

    public GraphicsElement removeChild(int index) {
        GraphicsElement removed = getChildren().remove(index);
        removed.setParent(null);
        return removed;
    }

    public int getChildIndex(GraphicsElement child) {
        if (isChild(child)) {
            return children.indexOf(child);
        } else {
            return -1;
        }
    }

    public void replace(GraphicsElement old, GraphicsElement newE) {
        int index = getChildIndex(old);

        if (index != -1) addChild(index, newE);
        else addChild(newE);
    }

    public GraphicsElement removeChild(GraphicsElement child) {
        int index = getChildren().indexOf(child);
        if (index >= 0) {
            return removeChild(index);
        } else {
            throw new IllegalArgumentException("The passed GraphicsElement is not a child of this GraphicsElement.");
        }
    }

    public boolean isChild(GraphicsElement child) {
        return (child != null && equals(child.getParent()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        GraphicsElement that = (GraphicsElement) o;

        return id == that.id;
    }

    @Override
    public int hashCode() {
        return id;
    }

    public GraphicsElement getParent() {
        return parent;
    }

    public void setParent(GraphicsElement parent) {
        this.parent = parent;
    }

    public boolean hasParent() {
        return getParent() != null;
    }

    @Override
    public boolean isActive() {
        return active;
    }

    @Override
    public void setActive(boolean active) {
        this.active = active;
    }

    @Override
    public MouseOptionMap getMouseOptionMap() {
        return mouseOptionMap;
    }

    @Override
    public void activate() {
        MouseEventListener.super.activate();
        children.forEach(GraphicsElement::activate);
    }

    @Override
    public void deactivate() {
        MouseEventListener.super.deactivate();
        children.forEach(GraphicsElement::deactivate);
    }

    public GApplet getApp() {
        return app;
    }

    public boolean isHidden() {
        return hidden;
    }

    public void hide() {
        hidden = true;
    }

    public void show() {
        hidden = false;
    }

    public int getHierarchyDepth() {
        return hierarchyDepth;
    }

    public int getId() {
        return id;
    }
}

File: GUILib/src/space/kiibou/gui/GraphicsElement.kt
package space.kiibou.gui

import javafx.beans.property.SimpleIntegerProperty
import javafx.beans.property.SimpleListProperty
import javafx.collections.FXCollections.observableArrayList
import javafx.collections.FXCollections.synchronizedObservableList
import processing.core.PConstants.*
import processing.core.PImage
import space.kiibou.GApplet
import space.kiibou.data.Rectangle
import space.kiibou.event.MouseEventConsumer
import space.kiibou.event.MouseEventListener
import space.kiibou.event.MouseEventOption
import space.kiibou.event.MouseOptionMap
import java.util.*
import kotlin.collections.ArrayList

val outline = System.getenv("outline")?.toBoolean() ?: false

abstract class GraphicsElement(val app: GApplet, x: Int = 0, y: Int = 0, width: Int = 0, height: Int = 0)
    : Rectangle(x, y, width, height), MouseEventListener {
    val scaleProp = SimpleIntegerProperty(1)
    val scale: Int
        get() = scaleProp.value
    val childrenProperty: SimpleListProperty<GraphicsElement> = SimpleListProperty(synchronizedObservableList(observableArrayList<GraphicsElement>()))
    val children: MutableList<GraphicsElement>
        get() = childrenProperty.value
    final override val mouseOptionMap: MouseOptionMap = MouseOptionMap()
    val id: Int = nextID()
    var hidden = false
        protected set
    private var parent: GraphicsElement? = null
    override var active: Boolean = true
    var hierarchyDepth: Int = 0
        private set
    private var preInitialized: Boolean = false
    private var initialized: Boolean = false
    private var postInitialized: Boolean = false
    private var insideDraw: Boolean = false
    private val deferredActions = Collections.synchronizedList(ArrayList<() -> Unit>())

    var clip: Boolean = true

    override fun registerCallback(option: MouseEventOption, callback: MouseEventConsumer) {
        super.registerCallback(option, callback)
        app.registerMethod("mouseEvent", this)
    }

    override fun unregisterCallback(option: MouseEventOption) {
        super.unregisterCallback(option)
        if (mouseOptionMap.isEmpty()) app.unregisterMethod("mouseEvent", this)
    }

    fun preInit() {
        preInitImpl()
        preInitialized = true
        children.forEach(GraphicsElement::preInit)
    }

    /**
     * Called before children are pre initialized
     */
    protected abstract fun preInitImpl()

    fun init() {
        children.forEach(GraphicsElement::init)
        initImpl()
        this.initialized = true
    }

    /**
     * Called after children are initialized
     */
    protected abstract fun initImpl()

    fun postInit() {
        children.forEach(GraphicsElement::postInit)
        postInitImpl()
        postInitialized = true
    }

    /**
     * Called before children are post initialized
     */
    protected abstract fun postInitImpl()

    fun draw() {
        if (!hidden) {
            insideDraw = true
            val g = app.graphics
            if (clip) {
                g.clip((x).toFloat(), (y).toFloat(), (width).toFloat(), (height).toFloat())
            }
            g.pushMatrix()
            g.pushStyle()
            drawImpl()
            children.forEach(GraphicsElement::draw)
            g.popStyle()
            g.popMatrix()
            if (clip) {
                g.noClip()
            }
            if (outline) {
                with(g) {
                    stroke(255f, 50f, 50f)
                    noFill()
                    rect(x.toFloat(), y.toFloat(), this@GraphicsElement.width.toFloat(), this@GraphicsElement.height.toFloat())
                    textSize(5f * scale)
                    textAlign(CENTER, TOP)
                    fill(50f, 155f, 50f)
                    text(unscaledWidth, x + this@GraphicsElement.width / 2f, y.toFloat())
                    textAlign(LEFT, CENTER)
                    text(unscaledHeight, x.toFloat(), y + this@GraphicsElement.height / 2f)
                }
            }
            insideDraw = false

            deferredActions.forEach { it() }
            deferredActions.clear()
        }
    }

    protected abstract fun drawImpl()

    fun deferAfterDraw(action: () -> Unit) {
        if (insideDraw) deferredActions.add(action)
        else action()
    }

    open fun moveTo(nx: Int, ny: Int): GraphicsElement {
        val dx = nx - x
        val dy = ny - y
        move(dx, dy)
        return this
    }

    open fun move(dx: Int, dy: Int): GraphicsElement {
        x = x.plus(dx)
        y = y.plus(dy)
        return this
    }

    fun resize(width: Int, height: Int): GraphicsElement {
        this.width = width
        this.height = height
        return this
    }

    fun resizeUnscaled(width: Int, height: Int): GraphicsElement {
        this.width = width * scale
        this.height = height * scale
        return this
    }

    val unscaledWidth: Int
        get() = width / scale

    val unscaledHeight: Int
        get() = height / scale

    operator fun plusAssign(element: GraphicsElement) {
        addChild(element)
    }

    operator fun plusAssign(p: Pair<Int, GraphicsElement>) {
        addChild(p.first, p.second)
    }

    open fun addChild(element: GraphicsElement) {
        addChild(children.size, element)
    }

    fun addChild(index: Int, element: GraphicsElement) {
        if (insideDraw)
            throw IllegalStateException("Can not modify children during draw")

        if (element.parent == null) {
            element.parent = this
            children.add(index, element)
            element.scaleProp.bind(scaleProp)
            element.hierarchyDepth = hierarchyDepth + 1
        } else {
            throw IllegalArgumentException("The passed GraphicsElement is already child of another GraphicsElement.")
        }
    }

    open operator fun get(index: Int): GraphicsElement? {
        return children[index]
    }

    open fun removeChild(index: Int): GraphicsElement {
        if (insideDraw)
            throw IllegalStateException("Can not modify children during draw")

        val removed: GraphicsElement = children.removeAt(index)
        removed.parent = null
        removed.scaleProp.unbind()
        return removed
    }

    fun getChildIndex(child: GraphicsElement?): Int = when {
        isChild(child) -> children.indexOf(child)
        else -> -1
    }

    fun replace(old: GraphicsElement?, newE: GraphicsElement) {
        if (insideDraw)
            throw IllegalStateException("Can not modify children during draw")

        val index = getChildIndex(old)
        if (index != -1) {
            removeChild(index)
            addChild(index, newE)
        } else addChild(newE)
    }

    fun removeChild(child: GraphicsElement?): GraphicsElement {
        if (insideDraw)
            throw IllegalStateException("Can not modify children during draw")

        val index = children.indexOf(child)
        return if (index >= 0) {
            removeChild(index)
        } else {
            throw IllegalArgumentException("The passed GraphicsElement is not a child of this GraphicsElement.")
        }
    }

    fun isChild(child: GraphicsElement?): Boolean {
        return child != null && equals(child.parent)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as GraphicsElement
        return id == that.id
    }

    final override fun toString(): String {
        return "${this.javaClass.simpleName}@${hashCode()}"
    }

    override fun hashCode(): Int {
        return id
    }

    fun hasParent(): Boolean {
        return parent != null
    }

    override fun activate() {
        super.activate()
        children.forEach(GraphicsElement::activate)
    }

    override fun deactivate() {
        super.deactivate()
        children.forEach(GraphicsElement::deactivate)
    }

    fun hide() {
        hidden = true
    }

    fun show() {
        hidden = false
    }
}

private var gID = 0

fun tilemapRenderFactory(image: PImage): (Int, Int, Int, Int) -> TileRenderer {
    return { x, y, w, h -> tileRenderFactory(x, y, x + w, y + h, image) }
}

private fun tileRenderFactory(u1: Int, v1: Int, u2: Int, v2: Int, tilemap: PImage): TileRenderer {
    return { g, x, y, w, h -> g.image(tilemap, x.toFloat(), y.toFloat(), w.toFloat(), h.toFloat(), u1, v1, u2, v2) }
}

private fun nextID(): Int {
    return gID++
}